### ISSUE TITLE
Socket error - timed out for ipv6 proxy

### ISSUE THREAD

**Issue by Lerbytech** [LonamiWebs/Telethon#322]
I use my own private proxy and trying to connect my client to proxy.
Proxy is valid. Tg account is also ok - I can see messages if try to connect without proxy.
When I try to connect I get socket timed out exception. I am new to python, so perhaps I am doing something wrong.  

I've seen several issues about ipv6 proxy support, but as far as i understand it was implemented successfully. I checked sources and see that logic for ipv6 is included.
Based on issues I've imported pysocks lib to cast proxy field (is it valid?), so my code is below:
Code:
`clientP = TelegramClient(phone_number, api_id, api_hash, proxy=(socks.SOCKS5, '5.188.215.240', 5574, 1, login, pass), timeout=timedelta(seconds=3))`

`clientP.connect()`

Still I get exception 
I'll appreciate any help although I understand that problem might be quite obvious.
Thanks in advance.

Stacktrace:

> Unable to display children:Error resolving variables Traceback (most recent call last):
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 813, in connect
    negotiate(self, dest_addr, dest_port)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 477, in _negotiate_SOCKS5
    CONNECT, dest_addr)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 503, in _SOCKS5_request
    chosen_auth = self._readall(reader, 2)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 304, in _readall
    d = file.read(count - len(data))
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\socket.py", line 586, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out

>During handling of the above exception, another exception occurred:

>Traceback (most recent call last):
  File "C:/Users/Little-Dany/PycharmProjects/TG2/main2.py", line 265, in test
    clientP.connect();
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\telegram_bare_client.py", line 187, in connect
    self._sender.connect()
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\mtproto_sender.py", line 46, in connect
    self.connection.connect(self.session.server_address, self.session.port)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\connection.py", line 79, in connect
    self.conn.connect(ip, port)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\extensions\tcp_client.py", line 49, in connect
    self._socket.connect(address)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 96, in wrapper
    return function(*args, **kwargs)
  File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 817, in connect
    raise GeneralProxyError("Socket error", error)
socks.GeneralProxyError: Socket error: timed out

>During handling of the above exception, another exception occurred:

>Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_resolver.py", line 238, in resolve
    return dict[key]
KeyError: 'Exception'

>During handling of the above exception, another exception occurred:

>Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_comm.py", line 1004, in do_it
    _typeName, valDict = pydevd_vars.resolve_compound_variable(self.thread_id, self.frame_id, self.scope, self.attributes)
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_vars.py", line 246, in resolve_compound_variable
    var = getVariable(thread_id, frame_id, scope, attrs)
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_vars.py", line 238, in getVariable
    var = resolver.resolve(var, k)
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_resolver.py", line 240, in resolve
    return getattr(dict, key)
AttributeError: 'dict' object has no attribute 'Exception'

**Comment by Lonami** [comment]
> it was implemented successfully.

I actually never got IPv6 to work. That's why the issue wanting more transport modes doesn't (or shouldn't have) the IPv6 checkbox ticked.

So the first error is the socket times out, and then PyCharm itself has an error too?
> pydevd_resolver.py", line 238, in resolve
return dict[key]
KeyError: 'Exception'

**Comment by Lerbytech** [comment]
Yes, PyCharm too, but I don't think that's issue core reason. At least I hope so)
I catch exception simply by hovering mouse over Exception and reading its value, maybe this method had affected:
`try:
     client = blabla
     client.connect()
except Exception:
`

About proxy - I've tried to make telegram desktop through proxy and it doesn't work. So the problem may be also in proxy, will need more investigation.

Still - that is sad that you never got ipv6 to work. From sources i've thought that everything was fine. What problems have you faced with?

**Comment by Lonami** [comment]
Not connecting at all.

**Comment by Lerbytech** [comment]
Situation update: yesterday I've tried to make official desktop app to work through proxy. It didn't worked, but several hourse later finally all worked fine.
Today I've tested my code again and have another stacktrace:

> 
> Unable to display children:Error resolving variables Traceback (most recent call last):
>   File "C:/Users/Little-Dany/PycharmProjects/TG2/main2.py", line 264, in test
>     clientP.connect()
>   File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\telegram_bare_client.py", line 187, in connect
>     self._sender.connect()
>   File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\mtproto_sender.py", line 46, in connect
>     self.connection.connect(self.session.server_address, self.session.port)
>   File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\connection.py", line 79, in connect
>     self.conn.connect(ip, port)
>   File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\extensions\tcp_client.py", line 49, in connect
>     self._socket.connect(address)
>   File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 96, in wrapper
>     return function(*args, **kwargs)
>   File "C:\Users\Little-Dany\AppData\Local\Programs\Python\Python36-32\lib\site-packages\socks.py", line 812, in connect
>     negotiate = self._proxy_negotiators[proxy_type]
> TypeError: unhashable type: 'dict'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_resolver.py", line 238, in resolve
>     return dict[key]
> KeyError: 'Exception'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_comm.py", line 1004, in do_it
>     _typeName, valDict = pydevd_vars.resolve_compound_variable(self.thread_id, self.frame_id, self.scope, self.attributes)
>   File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_vars.py", line 246, in resolve_compound_variable
>     var = getVariable(thread_id, frame_id, scope, attrs)
>   File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_vars.py", line 238, in getVariable
>     var = resolver.resolve(var, k)
>   File "C:\Program Files\JetBrains\PyCharm Community Edition 2017.2.3\helpers\pydev\_pydevd_bundle\pydevd_resolver.py", line 240, in resolve
>     return getattr(dict, key)
> AttributeError: 'dict' object has no attribute 'Exception'

This leads me to three conclusions:
1) As I don't see any connection-timed-out-like messages, Telegram yesterday did have some issues with ipv6.
2) I still have problems with library and they are not the same as for previous report
3) PyCharm role may be important

@Lonami , you mentioned that ipv6 and telethon were not connecting at all. 
At that time stacktrace was similar to mine?

I'll try to exclude pycharm and try again.

**Comment by Lonami** [comment]
> TypeError: unhashable type: 'dict'

Are you sure you're passing the right parameters to the socket? It's trying to do this:
```
negotiate = self._proxy_negotiators[proxy_type]
```

But the `proxy_type` may be wrong? I don't know, sorry, I didn't write the proxy support on Telethon and haven't used proxies enough to know what's going on.

**Comment by Lerbytech** [comment]
New report on situation.

I've found that one man had exact problems like me with ipv6 proxies: https://toster.ru/q/441590 (in russian, but google translate works well). He had modified PySocks library (link to github https://github.com/Man197/PySocks) and had sent me his modified version of telethon lib.
I was trying to make this stuff work for 2 days now and came to dead end.

In new telethon lib as far as i understand author had modified TelegramClient constructor adding flag `use_v6 = True` and if flag is True then lib uses predefined ipv6 address of telegram dc.
What has been changed in PySocks? Well, it is better for you to check lates commit on github. 

> tl;dr Added ipv6 proxy which works on ipv4 address (PROXY_TYPE_HTTP_4TO6 flag ?)

Aaaand... well, I think that whole thing had come far way further than I can handle now with my current python skills. As C# developer I'm struggling with some fundamental problems of misunderstanding of python workflow and demand for help.

Fist things first: When I've started all this, I had two private proxies: ipv6 and ipv4. 
Ipv4 is working good with native telethon lib, but ipv6... I think that there may be some issues with it as even curl return 407 error when I try to check my ip. 

I've decided to check my proxies just in case. I've bought ipv6 proxy from another supplier and installed Proxifier to put all os traffic through each of my proxies. 
All ipv6 are not working (checked on google chrome)
Ipv4 is ok  - I managed to login to my acc through this proxy so telegram had sent me notification with ipv4 ip address.

For now, I have hypothesis that my internet provider doesn't support ipv6. Unfortunately their technical support will be available only on Monday.

On the other hand, I am not sure that I was good in installing modified libs.
1. I created virtual environment with help of PyCharm (not sure if it uses python  libs for this).
2. I put PySocks folder in project folder and started setup.py.
3. It didn't worked well so I monkeypatched 
`import socks`
to
`from PySocks import socks`
(I know that this is crappy af, but I didn't managed to make it in more... civilised way.
4. After that wanted to start setup.py of improved telethon - but there is none of that.
I wanted to try to use telethon's origin setup.py, but its source code is not known to me so i don't know what to modify to make it work
(however, while I'm writing this strings I think that I wouldn't break anything important if i'll start this script. Still, i'm too tired and not sure If new problems in case of their appearanse would have relevant source)
5. After all that I've tried different combinations of proxy settings.
In TelegramClient constructor I've tried all options of socks type field when setting proxy. 
I faced only 3 resulting exceptions when tried to call client.connect:
1. 407 error
2. 502 error
3. socket connection timed out
4*. (several times I faced "TelegramClient is not object' , but I can't reproduce it no so i have no idea wtf was it)

and.. that's all))

I don't know what to do know. 
Does it my provider's fault?
Am I missing something?
Did I installed libs correctly?
Maybe my proxies are messed up?
All of this is not my level stuff. I have work to do and for now - using proxifier to manage connections is suitable for my client, so I don't have time to make it pretty.

@Lonami , I suggest you to officially pm to aforementioned guy who managed to make ipv6, get his code and consultation and perhaps make ipv6 finally god damn working. If you both will face problems in translation I can help you. 

If something is not clear - please ask. I've tired to reproduce all steps as clear as I could, but still could not notice something.
p/s Also sorry for mistakes - I translate from english to russian much more frequently than opposite way)

**Comment by Lonami** [comment]
Sorry to hear you've gone through so much trouble to get IPv6 proxies working. But from what you've said, it seems like IPv6 support is weird (not because Telethon doesn't support it yet). My advice is that you test the IPv6 proxy with something you know that will work, and once you know that the proxy works correctly, try to use it in the library. Because currently you're facing issues with both the proxy and the support for IPv6 in the library, and fixing two problems at once sounds harder than going one by one.

You say @Man197 modified Telethon to work with IPv6 correctly? Hopefully this ping brings his attention and he can help out support such protocol from mainstream.

**Comment by Lerbytech** [comment]
Thank you for your answer. I'll definetely do this when I'll solve current problems. I'll call techsupport on Monday and report here about ipv6 in their network. Proxifier says that there is good connection between me and proxy, but no ping back from internet to proxy. 

I was editing my comment when you've answered, so I duplicate it hear: I also hope that @Man197 will bring his attention to us. 
On https://toster.ru/q/441590 page if  you click on author's nick, you will find his tg for contact. Perhaps it is way faster method to ping him here. If you'll need help in translation - call me.

Also, a friend  of mine just reported that he also didn' managed to make telegram cli work with ipv6. But Ipv4  is ok.

thanks for all

**Comment by Man197** [comment]
Hello.
I really managed to use Telethon with an IPv6 proxy.
By modifying the PySocks library, I have successfully connected via HTTP proxy, the input IP for which is v4, and the output is v6. Fork is here (https://github.com/Man197/PySocks). Added a new proxy type PROXY_TYPE_HTTP_4TO6. The task was solved within the project, the customer needed HTTP proxies, so I did not use SOCKS.
In the Telethon library, I simply added the use_v6 flag to the constructor of the TelegramClient class, the True value of which leads to the use of the v6 address to connect to the Telegram servers. I found the address for the connection somewhere on the Internet, I do not remember where.
At the moment, I can not check the situation with the connection via v6 proxy, because there is not (as far as I know) free v6 proxy for the test.
If necessary, I can make a fork of the Telethon library with my changes, but they are very simple, literally 5 lines of code.
Sorry for possible problems with translation from Russian into English :)

**Comment by Lonami** [comment]
@Man197 thanks for the reply. You probably should make a pull request to @Anorov's repository so that everyone can benefit from your changes.

> I can make a fork of the Telethon library with my changes, but they are very simple, literally 5 lines of code.

That would be awesome, even if it's really simple, since I don't work with proxies at all myself :)

**Comment by Man197** [comment]
@Lonami ОК, I'll make a pull request in the near future. But this code is not well tested, so if something is wrong, I'm sorry :)

**Comment by windx13** [comment]
Good afternoon! I'm following this topic. But I see that there is still no solution, unfortunately. I really need support for the ipv6 proxy. Are there other solutions to this problem?

**Comment by Lonami** [comment]
@windx13 I pushed `v0.15.5` to PyPi which has @Man197's additions. Note that if your machine doesn't support it or doesn't have IPv6 configured properly, you won't be able to use it anyway, but the option is there. See #425 for more.

@Lerbytech (or anyone else) do you think this issue should be closed now?

**Comment by Lerbytech** [comment]
Wow, great! I'll check it on Sunday and tell you more. Am I wrong that know we can buy ipv6 proxy and use it?

**Comment by Lonami** [comment]
I don't know enough to answer this question, sorry.

**Comment by windx13** [comment]
Still not yet. Undoubtedly, a lot of work has been done - added support for the proxy. But, I today with Man197 tested the proxy. Connect failed due to the fact that you need to make some more changes to Telethon. Man197 promised to do in the near future a new Pull Requests.

**Comment by Lonami** [comment]
@windx13, @Man197 promised and did the pull request #425 which has been both merged and pushed to PyPi as `v0.15.5`. Unless you tested this and indeed more changes are needed. Then I don't know.

**Comment by Lerbytech** [comment]
I think that we need to use same settings for os to deal with ipv6 or at least exchange what we did.

My provider in campus doesn't support ipv6 afaik and I have only options to use mobile phone as router or to work from vds. I'll try to check both ways on weekend

**Comment by Man197** [comment]
I think we need to change the format of the use_ipv6 argument and bring it to the choice of one of the three flags: use_ipv4 OR use_ipv6 OR use_ipv6_with_AF_INET4_family. The fact is that the ipv6 proxy, for which all changes are started, fully encapsulate the work with ipv6, for the user of such a proxy the process looks as if it communicates on ipv4, but refers to the ipv6 address. In this case, you need the ipv6 address of the telegram server, but using the AF_INET4 socket. I will soon prepare a pull request.

**Comment by Lonami** [comment]
Thanks again @Man197. This surely looks like [another](https://github.com/LonamiWebs/Telethon/blob/976777414756bfc0e14a056cdd160e56089ed1ae/telethon/network/connection.py#L14-L34) good case for the [`enum`](https://docs.python.org/3/library/enum.html) module :)

**Comment by Lonami** [comment]
We can close this once `v2` lands with https://github.com/LonamiWebs/Telethon/commit/c1a40630a3477ad9b03b44e009ca57928de887d7, since we replace `pysocks` with `aiosocks`, and hopefully that solves it.

**Comment by Lonami** [comment]
Actually, let's close it already, since the commit has been made. It just hasn't been released. Then [the milestone](https://github.com/LonamiWebs/Telethon/milestone/1) will show progress nicely.

**Comment by Lerbytech** [comment]
Wow, I've long forgotten this thread. Even project which required this library is long gone )))
Still, impressed with your work and progres. It's cool that you still maintain and develop library and even take care about old issues. Congrats!
